### PR TITLE
Add basic dataset metadata editor UI

### PR DESCRIPTION
**Pull request by kulyk**
Adds basic dataset metadata editing UI on `/dataset/:id/metadata`

**Out of scope**

Metadata editing requires many changes, so the work is going to be broken down into a few PRs targeting `dataset/metadata-editor` feature branch. This PR is lacking a few things:

* API integration: it's not possible to save metadata changes
* field mapping UI for native datasets (for native datasets, people will need to explicitly set to which DB column a field is mapped to)
* semantic type input (this PR uses a simple "select" input, though we actually want to have a slightly customized AccordionList thing
* field formatting UI (e.g. once you set the field's semantic type to "Cost", you should be able to specify its currency, number formatting options, etc.)

### To Verify

1. Open a dataset with a structured query
2. Click the dataset name to open the details sidebar and click the new "Customize metadata" button
3. You should end up on the dataset editor page and:
    * you should see navigation tabs in the top bar, the "Metadata" tab should be selected
    * you should not see the query editor, only the results table and a sidebar on the right
    * the right sidebar should display field metadata: display name, field description, semantic type, etc.
    * you shouldn't be able to resize the results table
4. The sidebar should display the first field's metadata by default
5. Click on a few column names, they should get selected and the sidebar should now display their metadata
6. Ensure metadata matches the one on `/admin/datamodel`
7. Click the "Query" tab in the top bar, the editor should switch into the query editor mode. The metadata sidebar should disappear, and the query editor should work as described in #19103
7. Open the metadata editor for a native dataset. It should look pretty much the same, but the fields should be lacking most of the metadata (it's unclear from which table a column comes from, so we can't "inherit" the metadata by default and people will need to explicitly set to which real column a field is mapped to)

### Demo

https://user-images.githubusercontent.com/17258145/144880535-31eb5be3-67f9-4b41-b7e6-ca199f058be6.mov

### CODE REVIEW

**Comment by daltojohnso** [comment]
Screen Shot 2021-12-06 at 12 53 02 PM

I noticed there's quite a delay when you click on a column header in the "edit metadata" mode. Looks like a chunk of the slowness is caused by a hook in `SyncSnackbar.jsx`. Not sure what that's about.

**Comment by daltojohnso** [comment]
https://user-images.githubusercontent.com/13057258/144928371-60697733-f75b-4fd3-aca3-cec8e1ccb568.mov

Also, perhaps related to the above performance problem I commented on, but it seems like the form ui on the right side is pretty slow, too. Every key press in the description input is ~60ms long and you can feel the sluggishness. Seems like some unnecessary rendering going on. Didn't dig too deep into it, however.

**Review by daltojohnso — CHANGES_REQUESTED** [review]
One minor thing besides some comments regarding performance -- feel free to look at performance stuff in a subsequent PR. Besides that, I noticed that custom columns don't seem to work:

https://user-images.githubusercontent.com/13057258/144929992-eb5cfcdb-3b20-4557-bd7b-5a45e17995a7.mov

**Comment by mazameli** [comment]
We shouldn't discard metadata changes when you tab between the query and metadata modes: https://www.loom.com/share/de53f08706cd47d9ac435e92a9ea367d

**Comment by mazameli** [comment]
> I noticed there's quite a delay when you click on a column header in the "edit metadata" mode. Looks like a chunk of the slowness is caused by a hook in SyncSnackbar.jsx. Not sure what that's about.

@daltojohnso `SyncSnackbar.jsx` is a new component that the Activation team added. It shows up in the bottom right of the app after you connect a new DB to let you know if it's done syncing yet. cc @ranquild

**Comment by mazameli** [comment]
I'm just assuming the specced tabbing behavior [isn't implemented yet](https://www.loom.com/share/d51f619f79d8442c85befb723c9e77b8).

**Comment by mazameli** [comment]
Is it expected that I can't yet map a foreign key field to the column it points to in the related table? https://www.loom.com/share/10a8f510f2f94b09bb9b16196964f737

**Review by mazameli — CHANGES_REQUESTED** [review]
Presuming this PR isn't meant to persist/save the metadata changes you make, but just add the UI, then it's looking really cool. 

Just one small chage request: the space between column headings is too tight currenty. See screenshots. 

**Desired spacing between column headings: 24px**

image

**Current spacing is too tight**

image

**Comment by mazameli** [comment]
Just noticed this, too: if I cancel my edits, my results disappear: https://www.loom.com/share/01c25e079a954a7c81c21328526eef47

**Comment by kulyk** [comment]
@daltojohnso many thanks for the performance investigation, the videos and screenshots helped a lot

I did a few things and looks like the performance has improved (e.g. key press handling in metadata sidebar now takes ~20-30ms compared to 60ms before the optimization)

* query editor should not rerender now when it's hidden in "metadata" mode ([commit](url))
* some memorization in the [sidebar component](https://github.com/metabase/metabase/pull/19206/commits/0521206e24725c4fc305e647e4086973e473dec3) and the [dataset editor itself](https://github.com/metabase/metabase/pull/19206/commits/eede1b8f57f69565ffa95be649ba6eaa7adabc0b)

File: frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx [hunk 1]
```diff
@@ -41,8 +52,27 @@ const propTypes = {
 };
 
 const INITIAL_NOTEBOOK_EDITOR_HEIGHT = 500;
+const TABLE_HEADER_HEIGHT = 45;
+
+const isSameField = (f1, f2) => {
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx]: Tried that and it seems to work good. Thanks!

**Comment by kulyk** [comment]
@mazameli thanks for your review

> It seems as though this PR isn't meant to actually persist/save the metadata yet (right?)

Correct, the API isn't available yet

> I'm just assuming the specced tabbing behavior [isn't implemented yet](https://www.loom.com/share/d51f619f79d8442c85befb723c9e77b8), and will be in a different PR, which is cool by me.

Correct

> Is it expected that I can't yet map a foreign key field to the column it points to in the related table?

Yes. If you don't mind, I'd batch that together with field mapping for native datasets (working on that now, should be the next PR).

> Just one small change request: the space between column headings is too tight currently

Thanks! I also noticed it wasn't properly handling long column names. Both issues should be fixed 🙌 

> if I cancel my edits, my results disappear

Yeah, Jeff caught that during query editor bug bash. The issue is tracked, I'd prefer to issue the fix in another PR

**Review by mazameli — APPROVED** [review]

**Review by daltojohnso — APPROVED** [review]

**Pull request merged**